ext/dom/element.c

```diff
@@ -858,8 +960,54 @@ PHP_METHOD(DOMElement, setAttributeNS)
 	if (errorcode != 0) {
 		php_dom_throw_error(errorcode, stricterror);
 	}
+}
 
-	RETURN_NULL();
+/* https://dom.spec.whatwg.org/#dom-element-setattributens */
+static void dom_set_attribute_ns_modern(dom_object *intern, xmlNodePtr elemp, zend_string *uri, zend_string *name, const char *value)
+{
+	xmlChar *localname = NULL, *prefix = NULL;
+	int errorcode = dom_validate_and_extract(uri, name, &localname, &prefix);
+
+	if (errorcode == 0) {
+		php_dom_libxml_ns_mapper *ns_mapper = php_dom_get_ns_mapper(intern);
+		xmlNsPtr ns = php_dom_libxml_ns_mapper_get_ns_raw_prefix_string(ns_mapper, prefix, xmlStrlen(prefix), uri);
+		if (UNEXPECTED(xmlSetNsProp(elemp, ns, localname, BAD_CAST value) == NULL)) {
+			php_dom_throw_error(INVALID_STATE_ERR, /* strict */ true);
+		}
+	} else {
+		php_dom_throw_error(errorcode, /* strict */ true);
+	}
+
+	xmlFree(localname);
+	xmlFree(prefix);
+}
+
+/* {{{ URL: http://www.w3.org/TR/2003/WD-DOM-Level-3-Core-20030226/DOM3-Core.html#core-ID-ElSetAttrNS
+Modern spec URL: https://dom.spec.whatwg.org/#dom-element-setattributens
+Since: DOM Level 2
+*/
+PHP_METHOD(DOMElement, setAttributeNS)
+{
+	zval *id;
+	xmlNodePtr elemp;
+	size_t value_len = 0;
+	char *value;
+	zend_string *uri;
+	zend_string *name = NULL;
+	dom_object *intern;
+
+	id = ZEND_THIS;
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "S!Ss", &uri, &name, &value, &value_len) == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	DOM_GET_OBJ(elemp, id, xmlNodePtr, intern);
+
+	if (php_dom_follow_spec_intern(intern)) {
+		dom_set_attribute_ns_modern(intern, elemp, uri, name, value);
+	} else {
+		dom_set_attribute_ns_legacy(intern, elemp, uri ? ZSTR_VAL(uri) : NULL, ZSTR_VAL(name), value, uri ? ZSTR_LEN(uri) : 0, ZSTR_LEN(name));
+	}
```

Comment:
I do think passing zend_string to the legacy function probably makes sense.

Reply:
Maybe, maybe not. Having value and length separately makes it easy to have a (NULL, 0) pair for char* and length. I tried to quickly update this API here, but there's a few more APIs down the line that need to change; and I don't want to refactor this in this large PR with a risk of breakage.

Reply:
ACK